src/transformers/models/falcon/modeling_falcon.py

```diff
@@ -416,7 +639,11 @@ def __init__(self, config: FalconConfig):
         super().__init__()
         hidden_size = config.hidden_size
         self.num_heads = config.num_attention_heads
-        self.self_attention = FalconAttention(config)
+        self.self_attention = (
+            FalconAttention(config)
+            if not getattr(config, "_use_flash_attn_2", False)
```

Comment:
Do we really want to control whether flash attention should be used or not via a config variable?
This goes IMO a bit against the libraries design in that config files should only store information about the model architecture not whether "optimizations" should be enabled or not. E.g. we removed `gradient_checkpointing` from the config and it's **always** disabled by default now. The user can enable it if needed, but it shouldn't be stored in the config.

 Also I cannot find the config variable anywhere in the current codebase.

Reply:
I think the main issue with that is when someone saves a model and loads it back without flash attention, there it might cause issues. What about removing that attribute when saving it similarly as other attributes: https://github.com/huggingface/transformers/blob/main/src/transformers/configuration_utils.py#L870-L873 ? That way the only possible way to enable FA2 is to add `use_flash_attn_2` in `from_pretrained` or create a config with the attribute `_use_flash_attn_2` which should be discouraged as that attribute is private, unless users know what they're doing.
Currently instantiating the flashattention modules at init is the simplest and safest way, especially for models loaded with accelerate, or quantized models.